<h2 style="text-align:center;"> Teguh Ryan - 1203230044 <h2>

RUMUS ECLUDIAN

akar((x2-x1)kuadrat + (y2-y1)kuadrat)

Import library

In [23]:
import random
import math

Deklarasi function

In [51]:
distances = {
    'A': {'A': 0, 'B': 10, 'C': 15, 'D': 20, 'E': 25},
    'B': {'A': 10, 'B': 0, 'C': 35, 'D': 25, 'E': 15},
    'C': {'A': 15, 'B': 35, 'C': 0, 'D': 30, 'E': 20},
    'D': {'A': 20, 'B': 25, 'C': 30, 'D': 0, 'E': 10},
    'E': {'A': 25, 'B': 15, 'C': 20, 'D': 10, 'E': 0}
}

cities = ['A', 'B', 'C', 'D', 'E']


def total_distance(tour):
    dist = 0
    for i in range(len(tour)):
        dist += distances[tour[i]][tour[(i + 1) % len(tour)]]  
    return dist


def get_neighbors(tour):
    neighbors = []
    for i in range(len(tour)):
        for j in range(i + 1, len(tour)):
            neighbor = tour[:]
            neighbor[i], neighbor[j] = neighbor[j], neighbor[i] 
            neighbors.append(neighbor)
    return neighbors

current_solution = cities[:]  
random.shuffle(current_solution)  
current_distance = total_distance(current_solution)

print("Solusi awal (urutan kota):", current_solution)
print("Total jarak awal:", current_distance)

while True:
    neighbors = get_neighbors(current_solution)
    best_neighbor = min(neighbors, key=lambda tour: total_distance(tour))
    best_distance = total_distance(best_neighbor)
    
    if best_distance >= current_distance:
        break
    
    current_solution = best_neighbor
    current_distance = best_distance

print("\nSolusi terbaik (urutan kota):", current_solution)
print("Total jarak terbaik:", current_distance)

Solusi awal (urutan kota): ['E', 'A', 'B', 'C', 'D']
Total jarak awal: 110

Solusi terbaik (urutan kota): ['C', 'A', 'B', 'E', 'D']
Total jarak terbaik: 80


1. Representasi solusi awal

    Solusi awal: E → A → B → C → D → E

    Hitung total jarak rute ini berdasarkan tabel:

        E → A = 10

        A → B = 35

        B → C = 30

        D → E = 10

        E → A = 25

    Total jarak: 10 + 35 + 30 + 10 + 25 = 110

2. Cara menghasilkan neighbors

    jika solusi awal adalah:

    A → B → C → D → E → A

    Maka kita dapat menghasilkan neighbor dengan menukar dua kota di tengah, seperti:

        Tukar A dan B → E → B → A → C → D → E

        Tukar A dan C → E → C → B → A → D → E

        Tukar A dan D → E → D → B → C → A → E

        Tukar B dan C → E → A → C → B → D → E

        Tukar B dan D → E → A → D → C → B → E

        Tukar C dan D → E → A → B → D → C → E

    Setiap kombinasi pertukaran dua kota menghasilkan satu neighbor baru. Proses ini dilakukan untuk mengeksplorasi kemungkinan rute yang lebih pendek dari solusi saat ini.

3. Gunakan algoritma Hill Climbing untuk mencari rute yang paling pendek
    Langkah-langkah algoritma Hill Climbing:

    Solusi awal: E → A → B → C → D → E
    Hitung total jaraknya berdasarkan tabel:

        E → A = 25

        A → B = 10

        B → C = 35

        C → D = 30

        D → E = 10

    Total jarak awal = 25 + 10 + 35 + 30 + 10 = 110
    Langkah 1: Bangkitkan neighbors dari solusi awal

    Dengan menukar dua kota di tengah (A, B, C, D), kita dapatkan 6 neighbor:

        E → B → A → C → D → E

        E → C → B → A → D → E

        E → D → B → C → A → E

        E → A → C → B → D → E

        E → A → D → C → B → E

        E → A → B → D → C → E

    Langkah 2: Hitung jarak tiap neighbor

        E → B → A → C → D → E
        15 + 10 + 15 + 30 + 10 = 80 ✅ (lebih baik)

        E → C → B → A → D → E
        20 + 35 + 10 + 20 + 10 = 95

        E → D → B → C → A → E
        10 + 25 + 35 + 15 + 25 = 110

        E → A → C → B → D → E
        25 + 15 + 35 + 25 + 10 = 110

        E → A → D → C → B → E
        25 + 20 + 30 + 35 + 15 = 125

        E → A → B → D → C → E
        25 + 10 + 25 + 30 + 20 = 110

    Langkah 3: Pilih neighbor terbaik

    Rute dengan total jarak terpendek adalah 80, yaitu:

    E → B → A → C → D → E
    Langkah 4: Lanjutkan dari solusi baru

    Sekarang solusi baru adalah: E → B → A → C → D → E

    Cek neighbor dari rute ini, dengan menukar dua kota dari A, B, C, D (tetap tanpa E):

        E → A → B → C → D → E (kembali ke solusi awal, 110)

        E → C → A → B → D → E

        E → D → A → C → B → E

        Dst.

    Namun, tidak ada neighbor baru yang lebih baik dari 80, sehingga algoritma berhenti di sini.

4. Apa rute terpendek yang ditemukan oleh algoritma? Apakah rute tersebut merupakan solusi paling optimal atau masih merupakan local optimum?

    dikarenakan Rute terpendek yang ditemukan oleh algoritma : E → B → A → C → D → E = 15 + 10 + 15 + 30 + 10 = 80
    maka, solusi ini adalah local optimum, bukan jaminan solusi global terbaik (karena algoritma Hill Climbing hanya mencari solusi terbaik secara lokal.)